src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -20,538 +18,574 @@
-        selectionInfo = AdjustFinalTokensBasedOnContext(selectionInfo, model, cancellationToken);
-        selectionInfo = AssignFinalSpan(selectionInfo, text);
-        selectionInfo = ApplySpecialCases(selectionInfo, text, SemanticDocument.SyntaxTree.Options, _localFunction);
-        selectionInfo = CheckErrorCasesAndAppendDescriptions(selectionInfo, root);
-
-        // there was a fatal error that we couldn't even do negative preview, return error result
-        if (selectionInfo.Status.Failed)
-            return (null, selectionInfo.Status);
-
-        var controlFlowSpan = GetControlFlowSpan(selectionInfo);
-        if (!selectionInfo.SelectionInExpression)
+        private readonly bool _localFunction = localFunction;
+
+        protected override bool AreStatementsInSameContainer(StatementSyntax statement1, StatementSyntax statement2)
         {
-            var statementRange = GetStatementRangeContainedInSpan<StatementSyntax>(root, controlFlowSpan, cancellationToken);
-            if (statementRange == null)
-            {
-                selectionInfo = selectionInfo.WithStatus(s => s.With(succeeded: false, CSharpFeaturesResources.Cannot_determine_valid_range_of_statements_to_extract));
-                return (null, selectionInfo.Status);
-            }
+            if (statement1.Parent == statement2.Parent)
+                return true;
 
-            var isFinalSpanSemanticallyValid = IsFinalSpanSemanticallyValidSpan(model, controlFlowSpan, statementRange.Value, cancellationToken);
-            if (!isFinalSpanSemanticallyValid)
+            if (statement1.Parent is GlobalStatementSyntax
+                && statement2.Parent is GlobalStatementSyntax
+                && statement1.Parent.Parent == statement2.Parent.Parent)
             {
-                // check control flow only if we are extracting statement level, not expression
-                // level. you can not have goto that moves control out of scope in expression level
-                // (even in lambda)
-                selectionInfo = selectionInfo.WithStatus(s => s.With(succeeded: true, FeaturesResources.Not_all_code_paths_return));
+                return true;
             }
+
+            return false;
         }
 
-        var selectionChanged = selectionInfo.FirstTokenInOriginalSpan != selectionInfo.FirstTokenInFinalSpan || selectionInfo.LastTokenInOriginalSpan != selectionInfo.LastTokenInFinalSpan;
-
-        var result = await CSharpSelectionResult.CreateAsync(
-            selectionInfo.OriginalSpan,
-            selectionInfo.FinalSpan,
-            selectionInfo.SelectionInExpression,
-            doc,
-            selectionInfo.FirstTokenInFinalSpan,
-            selectionInfo.LastTokenInFinalSpan,
-            selectionChanged,
-            cancellationToken).ConfigureAwait(false);
-        return (result, selectionInfo.Status);
-    }
+        protected override SelectionInfo GetInitialSelectionInfo(CancellationToken cancellationToken)
```

Comment:
broke out smaller pieces of functionality so base class can share most of the lgoic between C# and VB.